Add hip related functions

rbramand-xilinx opened the pull request:
<!-- Thanks for sending a pull request! Please fill out below, remove sections that don't apply for your pull request.  -->
#### Problem solved by the commit
Added new hip module function as well as common utility function for creating host buffer

#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
#### How problem was solved, alternative solutions (if any) and why they were rejected


#### Risks (if any) associated the changes in the commit
Low

#### What has been tested and how, request additional testing if necessary
Tested build flow on Linux and Windows
Ran xrt_flow test on Linux and it passes

#### Documentation impact (if any)
NA

rbramand-xilinx commented:
This PR contains changes from other PR (https://github.com/Xilinx/XRT/pull/8110) which is open so added do not merge tag.
Once that PR gets merged I will rebase this and remove the tag.

stsoe commented on src/runtime_src/hip/core/module.cpp:
```diff
@@ -25,13 +25,13 @@ module(std::shared_ptr<context> ctx, const std::string& file_name)
   // it as xclbin data. Application can crash/seg fault
   // when improper data is passed
-  m_xclbin = xrt::xclbin{static_cast<axlf*>(image)};
+  m_xclbin = xrt::xclbin{static_cast<axlf*>(const_cast<void*>(image))};
```
Why this const_cast?  xclbin ctor takes const axlf*

stsoe reviewed (changes requested):
I don't think we should copy xrt_flow into the xrt repo.  This kind of test case belongs under testcases., or should be much simplified.     Copying a boatload of code means two places to maintain same boatload of code, not ideal.  We did the same mistakenly in XRT-MCDM for an even simpler dputest, and today that test doesn't work and no one has the energy to check and fix it.

rbramand-xilinx commented:
Hi @stsoe, updated the PR to remove xrt_flow test and I will push this test to testcases repo.
Updated the PR description appropriately. Please review

stsoe commented on tests/hip/common/common.h:
```diff
@@ -116,7 +116,34 @@ hip_test_device_bo {
+  T *_buffer;
+
+public:
+  explicit hip_test_host_bo(size_t size, unsigned int flags) : _buffer(nullptr) {
```
Not needed in this, doesn't hurt either.
```suggestion
  hip_test_host_bo(size_t size, unsigned int flags) : _buffer(nullptr) {
```

stsoe commented on tests/hip/common/common.h:
```diff
@@ -116,7 +116,34 @@ hip_test_device_bo {
+    test_hip_check(hipHostMalloc((void**)&_buffer, size * sizeof(T), flags));
+  }
+
+  ~hip_test_host_bo() noexcept {
+    test_hip_check(hipHostFree(_buffer));
+  }
```
Dtors are automatically noexcept.  The program will crash if `test_hip_check` throws.  If the dtor executes any code that can throw, then it would be better to try/catch and attempt to exit grace fully.
```suggestion
  ~hip_test_host_bo()  {
    test_hip_check(hipHostFree(_buffer));
  }
```

stsoe reviewed (approved):
Looks good.  Minor comments.
